File: src/event-cycle.js
import { Rational } from "./exact.js";

// Event-defined cycles deliberately contain occurrences rather than a mean
// duration.  `at` is an exact coordinate in `period.frame`; a boundary may
// also name the observed event that established that coordinate.  The finite
// series is the authority: there is no extrapolation before its first or
// after its last boundary.
function malformed(message) {
  return { valid: false, error: message, boundaries: [] };
}

export function eventBoundarySeries(period) {
  if (period?.kind !== "event-defined") return malformed("not an event-defined period");
  if (typeof period.frame !== "string" || !period.frame) return malformed("event-defined period needs a boundary frame");
  if (!Array.isArray(period.boundaries) || period.boundaries.length < 2) {
    return malformed("event-defined period needs at least two explicit boundaries");
  }
  const ids = new Set();
  const boundaries = [];
  for (const [index, source] of period.boundaries.entries()) {
    const id = String(source?.id || "").trim();
    if (!id) return malformed(`boundary ${index + 1} needs an id`);
    if (ids.has(id)) return malformed(`boundary ids must be unique (${id})`);
    ids.add(id);
    try {
      const at = Rational.parse(source?.at);
      if (boundaries.length && at.compare(boundaries.at(-1).at) <= 0) {
        return malformed("event-defined boundaries must be strictly ordered and unambiguous");
      }
      boundaries.push({ id, at, ...(typeof source.event === "string" ? { event: source.event } : {}) });
    } catch {
      return malformed(`boundary ${id} needs an exact finite coordinate`);
    }
  }
  return { valid: true, frame: period.frame, boundaries };
}

/** Resolve the one authored interval containing focus, using [start, end). */
export function resolveEventCycle(period, focus) {
  const series = eventBoundarySeries(period);
  if (!series.valid) return { ...series, resolved: false };
  let at;
  try { at = Rational.parse(focus); } catch { return { ...series, resolved: false, error: "focus must be an exact coordinate" }; }
  let index = -1;
  for (let cursor = 0; cursor < series.boundaries.length - 1; cursor += 1) {
    const start = series.boundaries[cursor].at;
    const end = series.boundaries[cursor + 1].at;
    if (at.compare(start) >= 0 && at.compare(end) < 0) { index = cursor; break; }
  }
  if (index < 0) return { ...series, resolved: false, error: "focus is outside the authored boundary window" };
  const boundary = series.boundaries[index];
  const nextBoundary = series.boundaries[index + 1];
  return { ...series, resolved: true, index, boundary, nextBoundary, start: boundary.at, end: nextBoundary.at, period: nextBoundary.at.sub(boundary.at) };
}

/** Move by authored intervals.  Reverse traversal never infers a predecessor. */
export function stepEventCycle(period, focus, direction = 1) {
  const current = resolveEventCycle(period, focus);
  if (!current.resolved) return current;
  const delta = Number(direction);
  if (!Number.isInteger(delta) || delta === 0) return { ...current, resolved: false, error: "cycle step must be a non-zero whole number" };
  const index = current.index + delta;
  if (index < 0 || index >= current.boundaries.length - 1) {
    return { ...current, resolved: false, error: "requested cycle is outside the authored boundary window" };
  }
  const boundary = current.boundaries[index];
  const nextBoundary = current.boundaries[index + 1];
  return { ...current, resolved: true, index, boundary, nextBoundary, start: boundary.at, end: nextBoundary.at, period: nextBoundary.at.sub(boundary.at) };
}

/** A finite radial/spiral window made only from authored adjacent intervals. */
export function eventCycleWindow(period, focus, past = 0, future = 0) {
  const current = resolveEventCycle(period, focus);
  if (!current.resolved) return current;
  const before = Math.max(0, Math.floor(Number(past) || 0));
  const after = Math.max(0, Math.floor(Number(future) || 0));
  const first = current.index - before;
  const last = current.index + after + 1;
  if (first < 0 || last >= current.boundaries.length) {
    return { ...current, resolved: false, error: "requested window exceeds authored boundaries" };
  }
  return { ...current, windowStart: current.boundaries[first].at, windowEnd: current.boundaries[last].at, firstIndex: first, lastIndex: last - 1 };
}

File: app/lib/core/event_cycle.dart
// Event-defined cycles.
//
// A cycle whose period is OBSERVED rather than declared contains occurrences, not
// a mean duration: `at` is an exact coordinate in the period's own frame, and a
// boundary may name the event whose observation established it.
//
// THE FINITE AUTHORED SERIES IS THE AUTHORITY. There is no extrapolation before
// the first boundary or after the last, and no averaging in either direction --
// asked for a cycle outside the authored window, this refuses. An observed lunar
// month is 29.31 days here and 29.61 there, and replacing either with the mean
// synodic month would be inventing an observation nobody made.
//
// Refusals travel as the coordinate-law layer's own [Attempt], so "resolved, or
// the author's sentence about why not" has one shape across the model.

import 'coordinate_law.dart';
import 'eras.dart';
import 'exact.dart';
import 'records.dart';

/// One authored boundary: its own id, the exact coordinate it sits at, and the
/// observed event that established it when one is named.
typedef Boundary = ({String id, Rational at, String? event});

/// The authored series, and whose units its coordinates are in.
typedef BoundarySeries = ({String frame, List<Boundary> boundaries});

/// One authored interval, `[start, end)`. Its period is the SUBTRACTION of two
/// observations, never a mean.
class EventCycle {
  const EventCycle(this.series, this.index);

  final BoundarySeries series;
  final int index;

  Boundary get boundary => series.boundaries[index];
  Boundary get next => series.boundaries[index + 1];
  Rational get start => boundary.at;
  Rational get end => next.at;
  Rational get period => end - start;
}

/// A window made only from authored ADJACENT intervals, for the radial and
/// spiral lenses.
typedef CycleWindow = ({
  EventCycle cycle,
  Rational start,
  Rational end,
  int firstIndex,
  int lastIndex,
});

Rational _exact(Object? value) => value is Rational ? value : Rational.parse(declaredText(value));

Attempt<BoundarySeries> eventBoundarySeries(Json? period) {
  if (declaredText(period?['kind']) != 'event-defined') {
    return const Refused('not an event-defined period');
  }
  final frame = declaredText(period?['frame']);
  if (frame.isEmpty) {
    return const Refused('event-defined period needs a boundary frame');
  }
  final rows = asList(period!['boundaries']);
  if (rows.length < 2) {
    return const Refused('event-defined period needs at least two explicit boundaries');
  }
  final ids = <String>{};
  final boundaries = <Boundary>[];
  for (final (index, row) in rows.indexed) {
    final entry = asMap(row);
    final id = declaredText(entry?['id']);
    if (id.isEmpty) return Refused('boundary ${index + 1} needs an id');
    if (!ids.add(id)) return Refused('boundary ids must be unique ($id)');
    final Rational at;
    try {
      at = _exact(entry?['at']);
    } catch (_) {
      return Refused('boundary $id needs an exact finite coordinate');
    }
    if (boundaries.isNotEmpty && at <= boundaries.last.at) {
      return const Refused('event-defined boundaries must be strictly ordered and unambiguous');
    }
    boundaries.add((
      id: id,
      at: at,
      event: entry?['event'] is String ? entry!['event'] as String : null,
    ));
  }
  return Resolved((frame: frame, boundaries: boundaries));
}

/// The one authored interval containing [focus], using `[start, end)`.
Attempt<EventCycle> resolveEventCycle(Json? period, Object? focus) {
  final series = eventBoundarySeries(period);
  final resolved = series.resolved;
  if (resolved == null) return Refused(series.refusal!);
  final Rational at;
  try {
    at = _exact(focus);
  } catch (_) {
    return const Refused('focus must be an exact coordinate');
  }
  final boundaries = resolved.boundaries;
  for (var cursor = 0; cursor < boundaries.length - 1; cursor += 1) {
    if (at >= boundaries[cursor].at && at < boundaries[cursor + 1].at) {
      return Resolved(EventCycle(resolved, cursor));
    }
  }
  return const Refused('focus is outside the authored boundary window');
}

/// Move by authored intervals. Reverse traversal never infers a predecessor, and
/// stepping past either end of the series refuses rather than extrapolating.
Attempt<EventCycle> stepCycle(EventCycle cycle, int direction) {
  if (direction == 0) {
    return const Refused('cycle step must be a non-zero whole number');
  }
  final index = cycle.index + direction;
  if (index < 0 || index >= cycle.series.boundaries.length - 1) {
    return const Refused('requested cycle is outside the authored boundary window');
  }
  return Resolved(EventCycle(cycle.series, index));
}

Attempt<EventCycle> stepEventCycle(Json? period, Object? focus, [int direction = 1]) {
  final current = resolveEventCycle(period, focus);
  final cycle = current.resolved;
  return cycle == null ? current : stepCycle(cycle, direction);
}

/// A finite radial/spiral window: [past] intervals before the focus and [future]
/// after it, refused outright when the authored series does not reach that far.
Attempt<CycleWindow> eventCycleWindow(Json? period, Object? focus, {int past = 0, int future = 0}) {
  final current = resolveEventCycle(period, focus);
  final cycle = current.resolved;
  if (cycle == null) return Refused(current.refusal!);
  final first = cycle.index - (past < 0 ? 0 : past);
  final last = cycle.index + (future < 0 ? 0 : future) + 1;
  if (first < 0 || last >= cycle.series.boundaries.length) {
    return const Refused('requested window exceeds authored boundaries');
  }
  return Resolved((
    cycle: cycle,
    start: cycle.series.boundaries[first].at,
    end: cycle.series.boundaries[last].at,
    firstIndex: first,
    lastIndex: last - 1,
  ));
}
